FernFlower-Patches/0041-More-correctly-handle-inner-class-synthetic-paramete.patch

```diff
@@ -0,0 +1,155 @@
+-
+-        for (int i = 0; i < interPairMask.size(); i++) {
+-          if (interPairMask.get(i) != null) {
+-            if (found) {
+-              interPairMask.set(i, null);
+-            }
+-            found = true;
+-          }
+-        }
++      } else {
++        mergeListSignatures(interPairMask, interMask, true);
+       }
+ 
+-      mergeListSignatures(interPairMask, interMask, true);
+-
+       for (VarFieldPair pair : interPairMask) {
+-        if (pair != null && !pair.fieldKey.isEmpty()) {
++        if (pair != null && !pair.fieldKey.isEmpty() && !NestedClassProcessor.ENUM_SYNTHETIC_SUPER.equals(pair.fieldKey)) {
+           nestedNode.mapFieldsToVars.put(pair.fieldKey, pair.varPair);
+         }
+       }
+@@ -764,7 +754,14 @@ public class NestedClassProcessor {
+           List<VarFieldPair> fields = new ArrayList<>(md.params.length);
+ 
+           int varIndex = 1;
+-          for (int i = 0; i < md.params.length; i++) {  // no static methods allowed
++          int start = 0;
++          if ((wrapper.getClassStruct().getAccessFlags() & CodeConstants.ACC_ENUM) != 0) { // enums have 2x synthetic params. not handled by normal detection because they go to the superclass constructor rather than to synthetic fields
++            fields.add(new VarFieldPair(NestedClassProcessor.ENUM_SYNTHETIC_SUPER, new VarVersionPair(-1, 0)));
++            fields.add(new VarFieldPair(NestedClassProcessor.ENUM_SYNTHETIC_SUPER, new VarVersionPair(-1, 0)));
++            varIndex += 2;
++            start += 2;
++          }
++          for (int i = start; i < md.params.length; i++) {  // no static methods allowed
+             String keyField = getEnclosingVarField(cl, method, graph, varIndex);
+             fields.add(keyField == null ? null : new VarFieldPair(keyField, new VarVersionPair(-1, 0))); // TODO: null?
+             varIndex += md.params[i].stackSize;
+@@ -819,17 +816,31 @@ public class NestedClassProcessor {
+     return fd.getName().contains("$") && fd.hasModifier(CodeConstants.ACC_FINAL) && fd.hasModifier(CodeConstants.ACC_PRIVATE);
+   }
+ 
++  /**
++   * Merge two sets of method signatures.
++   *
++   * This ensures that synthetic parameters are properly captured when field writes may go through another constructor.
++   *
++   * @param first the list to modify
++   * @param second the list compared against
++   * @param both if the comparison list should also be modified to align
++   */
+   private static void mergeListSignatures(List<VarFieldPair> first, List<VarFieldPair> second, boolean both) {
+-    int i = 1;
++    // Upstream this merges from the end of the list forwards
++    // However, synthetic parameters are added at the beginnings of method signatures (at least usually)
++    // so we really want to start from the beginning
++    // if it ends up that this assumption is not true, we'd have to do some sort of fuzzy matching here, or earlier
++    // on traverse multiple steps in the graph in getEnclosingVarField
```

Comment:
As discussed in Discord, this assumption is incorrect.
Would need a deeper dive into the compiler to find all cases where synthetics are added to the beginning or end.